### DISCUSSION TITLE
Looking for feedback on a status system

### DISCUSSION THREAD

**Discussion by jschuh** [jschuh/klipper-macros#111] [Ideas]
I'm thinking about adding a status system, to handle LEDs and other longtail cases that otherwise would require their own special-purpose customization. This would be similar to the layer system, but instead of firing off gcode on layer changes they would be fired on printer status changes.

As a broad sketch of how this could work, there would be something like a `SET_STATUS_EVENT` macro which would accept a macro and a status label. The macro would then be called when the status changes to the specified version, with the old status and new status passed as arguments (which the macro can choose to ignore). And you'd configure all of these status events by just calling the event registrations from the `_km_options` gcode so it gets run at startup, which would look something like this:

```
[gcode_macro _km_options]
# Options variables declared here
gcode:
  GCODE_ON_STATUS STATUS=ready COMMAND=status_ready
  GCODE_ON_STATUS STATUS=printing COMMAND=status_printing
  GCODE_ON_STATUS STATUS=paused COMMAND=status_paused
  # And more lines like this for every status you want to handle...
```

I'm thinking of the following statuses (which I've mostly cribbed from other sources) I'd like to hear about any I might have missed:

- **ready** - Printer is ready to receive a job
- **filament_load** - Loading filament
- **filament_unload** - Unloading filament
- **bed_heating** - Waiting for the bed to reach target
- **chamber_heating** - Waiting for the chamber to reach target
- **extruder_heating** -  Waiting for the extruder to reach target
- **homing** - Homing any axis
- **leveling_gantry** - Performing quad gantry-leveling
- **calibrating_z** - Performing z-tilt adjustment
- **meshing** - Calibrating a bed mesh
- **purging** - Printing purge line
- **printing** - Actively printing
- **paused** - Print is paused
- **cancelling** - Print is being cancelled
- **complete** - Print completed
- **idle** - Printer idle-timeout fired

**Comment by fire1ce** [comment]
Sounds great! I would suggest adding the m300 (beep) to the options. i find it very useful when running a m600 in a middle of the print and im not in the same room.

**Reply by jschuh** [reply]
I should reiterate that the point of having a status system is specifically to avoid directly implementing things like beeps or LEDs in this project. The idea is that the system just runs the registered gcode commands when the status changes to the one specified. So, you could implement your own beeps with something like this:

```
[gcode_macro beep_on_complete]
gcode:
  {% for i in range(5) %}
    M300 P100
    G4 P200
  {% endfor %}

[gcode_macro _km_options]
# Options variables declared here
gcode:
  GCODE_ON_STATUS STATUS=complete COMMAND=beep_on_complete
  # And more lines like the one above, for every status you want to handle
```

**Reply by fire1ce** [reply]
Sounds even better 😄

**Comment by jschuh** [comment]
I have an implementation of the event system up on the [status branch](https://github.com/jschuh/klipper-macros/tree/status#print-status-events).
